src/api.ts

```diff
@@ -0,0 +1,249 @@
+}
+
+export type ConnectionEvents = {
+	connect: []
+	disconnect: []
+	error: [error: Error]
+}
+
+export class BasicCasparCGAPI extends EventEmitter<ConnectionEvents> {
+	private _connection: Connection
+	private _host: string
+	private _port: number
+
+	private readonly _useSequential: boolean
+	private _requestQueue: Array<InternalRequest> = []
+	private _timeoutTimer: NodeJS.Timer
+	private _timeoutTime: number
+
+	constructor(options?: Options) {
+		super()
+
+		this._host = options?.host || '127.0.0.1'
+		this._port = options?.port || 5250
+		this._useSequential = options?.useSequential || false
+
+		this._connection = new Connection(this._host, this._port)
+
+		this._connection.on('connect', () => {
+			this.emit('connect')
+			this._processQueue().catch((e) => this.emit('error', e))
+		})
+		this._connection.on('disconnect', () => this.emit('disconnect'))
+
+		this._connection.on('data', (response) => {
+			if (this._useSequential) {
+				const request = this._requestQueue.shift()
+
+				if (request) {
+					request.resolve(response)
+				}
+			} else {
+				const request = this._requestQueue.find((req) => req.requestId === response.reqId)
+
+				if (request) {
+					request.resolve(response)
+					this._requestQueue = this._requestQueue.filter((req) => req.requestId !== response.reqId)
+				}
+			}
+
+			this._processQueue().catch(() => null)
+		})
+
+		this._timeoutTime = options?.timeoutTime || 5000
+		this._timeoutTimer = setInterval(() => this._checkTimeouts(), this._timeoutTime)
+	}
+
+	get host(): string {
+		return this._host
+	}
+
+	set host(host: string) {
+		this._host = host
+		this._connection.changeConnection(this._host, this._port)
+	}
+
+	get port(): number {
+		return this._port
+	}
+
+	set port(port: number) {
+		this._port = port
+		this._connection.changeConnection(this._host, this._port)
+	}
+
+	/** Stops internal timers so that the class is ready for garbage disposal */
+	discard(): void {
+		clearInterval(this._timeoutTimer)
+	}
+
+	executeCommand(command: AMCPCommand): Request {
+		const reqId = Math.random().toString(35).slice(2, 7)
+
+		let outerResolve: InternalRequest['sentResolve'] = () => null
+		// outerReject: InternalRequest['sentReject'] = () => null
+		const s = new Promise<SendResult>((resolve) => {
+			outerResolve = resolve
+			// outerReject = reject
+		})
+
+		const internalRequest: InternalRequest = {
+			requestId: reqId,
+			command,
+
+			// stubs to be replaced
+			resolve: () => null,
+			reject: () => null,
+
+			processed: false,
+			sentResolve: outerResolve,
+			// sentReject: outerReject,
+
+			request: {
+				requestId: reqId,
+				cancel: () =>
+					this._requestQueue.splice(
+						this._requestQueue.findIndex((r) => r.requestId === reqId),
+						1
+					),
+
+				sent: s,
+			},
+		}
+
+		this._requestQueue.push(internalRequest)
+		this._processQueue().catch((e) => this.emit('error', e))
+
+		return internalRequest.request
+	}
+
+	private async _processQueue(): Promise<void> {
```

Comment:
Regarding setting the `response`:
Could we not wait for `sentOK` to be resolved before setting the `response`? Currentlythe user has to
```javascript
const request = await connection.play({...})
await request.sent
const response = await request.response
```
to be able to use the response, which is inconvenient.

Reply:
Well, the command isn't guaranteed to be sent immediately (especially in synchronous mode, where 1 command is sent at a time) so if we want to be able to cancel it we have to return a promise like that. 

However you can do
```ts
const request = await connection.play({...})
const response = await request.sent && await request.response

if (response) {
  // ccg replied
} else {
  // it was not sent
}}
```